jerry-core/ecma/base/ecma-lcache.c

```diff
@@ -79,241 +74,169 @@ ecma_lcache_init (void)
 ecma_lcache_invalidate_entry (ecma_lcache_hash_entry_t *entry_p) /**< entry to invalidate */
 {
   JERRY_ASSERT (entry_p != NULL);
   JERRY_ASSERT (entry_p->object_cp != ECMA_NULL_POINTER);
-
-  ecma_deref_object (ECMA_GET_NON_NULL_POINTER (ecma_object_t,
-                                                entry_p->object_cp));
+  JERRY_ASSERT (entry_p->prop_name_cp != ECMA_NULL_POINTER);
+  JERRY_ASSERT (entry_p->prop_p != NULL);
 
   entry_p->object_cp = ECMA_NULL_POINTER;
-  ecma_deref_ecma_string (ECMA_GET_NON_NULL_POINTER (ecma_string_t,
-                                                     entry_p->prop_name_cp));
-
-  if (entry_p->prop_p != NULL)
-  {
-    ecma_set_property_lcached (entry_p->prop_p, false);
-  }
+  ecma_set_property_lcached (entry_p->prop_p, false);
 } /* ecma_lcache_invalidate_entry */
-#endif /* !CONFIG_ECMA_LCACHE_DISABLE */
 
 /**
- * Invalidate all entries in LCache
- */
-void
-ecma_lcache_invalidate_all (void)
-{
-#ifndef CONFIG_ECMA_LCACHE_DISABLE
-  for (uint32_t row_index = 0; row_index < ECMA_LCACHE_HASH_ROWS_COUNT; row_index++)
-  {
-    for (uint32_t entry_index = 0; entry_index < ECMA_LCACHE_HASH_ROW_LENGTH; entry_index++)
-    {
-      if (ecma_lcache_hash_table[ row_index ][ entry_index ].object_cp != ECMA_NULL_POINTER)
-      {
-        ecma_lcache_invalidate_entry (&ecma_lcache_hash_table[ row_index ][ entry_index ]);
-      }
-    }
-  }
-#endif /* !CONFIG_ECMA_LCACHE_DISABLE */
-} /* ecma_lcache_invalidate_all */
-
-#ifndef CONFIG_ECMA_LCACHE_DISABLE
-/**
- * Invalidate entries of LCache's row that correspond to given (object, property) pair
+ * Compute the row index of object / property name pair
+ *
+ * @return row index
  */
-static void
-ecma_lcache_invalidate_row_for_object_property_pair (uint32_t row_index, /**< index of the row */
-                                                     unsigned int object_cp, /**< compressed pointer
-                                                                              *   to an object */
-                                                     ecma_property_t *property_p) /**< pointer to the
-                                                                                   *   object's property */
+static inline size_t __attr_always_inline___
+ecma_lcache_row_idx (jmem_cpointer_t object_cp, /**< compressed pointer to object */
+                     const ecma_string_t *prop_name_p) /**< proeprty name */
 {
-  for (uint32_t entry_index = 0; entry_index < ECMA_LCACHE_HASH_ROW_LENGTH; entry_index++)
-  {
-    if (ecma_lcache_hash_table[ row_index ][ entry_index ].object_cp == object_cp
-        && ecma_lcache_hash_table[ row_index ][ entry_index ].prop_p == property_p)
-    {
-      ecma_lcache_invalidate_entry (&ecma_lcache_hash_table[ row_index ][ entry_index ]);
-    }
-  }
-} /* ecma_lcache_invalidate_row_for_object_property_pair */
+  return (size_t) ((ecma_string_hash (prop_name_p) ^ object_cp) & (ECMA_LCACHE_HASH_ROWS_COUNT - 1));
```

Comment:
This is randomizing the hash with the object. Before this patch, objects with the same field names evicted each other:

    a[0] = 1; b[0] = 1; c[0] = 1; // c[0] evicted a[0] before. Not anymore

This happened frequently in access-fannkuch test. Now because the hash is xor-ed with the object, properties do not evict each other anymore.